File: lilac/src/main/java/com/hartwig/hmftools/lilac/candidates/AminoAcidFiltering.java
package com.hartwig.hmftools.lilac.candidates;

import com.google.common.collect.Lists;
import com.google.common.collect.Sets;
import com.hartwig.hmftools.lilac.SequenceCount;
import com.hartwig.hmftools.lilac.seq.HlaSequenceLoci;

import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

public class AminoAcidFiltering
{
    private final List<Integer> mAminoAcidBoundaries;

    public AminoAcidFiltering(final List<Integer> aminoAcidBoundaries)
    {
        mAminoAcidBoundaries = aminoAcidBoundaries;
    }

    public List<HlaSequenceLoci> aminoAcidCandidates(final List<HlaSequenceLoci> candidates, final SequenceCount aminoAcidCount)
    {
        List<HlaSequenceLoci> results = Lists.newArrayList();
        results.addAll(candidates);

        Set<Integer> locations = Sets.newHashSet();

        for(int loci = 0; loci < aminoAcidCount.getLength(); ++loci)
        {
            if(mAminoAcidBoundaries.contains(loci))
                continue;

            // int depth = aminoAcidCount.depth(loci);

            List<String> expectedSequences = aminoAcidCount.getMinCountSequences(loci);

            final int lociConst = loci;

            results = results.stream()
                    .filter(x -> x.consistentWithAny(expectedSequences, Lists.newArrayList(lociConst)))
                    .collect(Collectors.toList());
        }

        return results;
    }

}

File: lilac/src/main/kotlin/com/hartwig/hmftools/lilac/candidates/AminoAcidFiltering.kt
package com.hartwig.hmftools.lilac.candidates

import com.hartwig.hmftools.lilac.SequenceCount
import com.hartwig.hmftools.lilac.seq.HlaSequenceLoci

class AminoAcidFiltering(private val aminoAcidBoundaries: Set<Int>) {

    fun aminoAcidCandidates(candidates: List<HlaSequenceLoci>, aminoAcidCount: SequenceCount): List<HlaSequenceLoci> {
        var result = candidates
        val locations = (0 until aminoAcidCount.length).toSet() subtract aminoAcidBoundaries
        for (loci: Int in locations) {
            //TODO: charles?
//            val depth = aminoAcidCount.depth(loci)
//            if (depth >= config.someParameter) {
                val expectedSequences = aminoAcidCount.sequenceAt(loci)
                result = result.filter { it.consistentWithAny(expectedSequences, loci) }
//            }
        }
        return result
    }



}